File: exercises/py110/medium2/triangle_sides.py
'''
Input:  3 ints representing sides of a triangle
Output: A string representing the type of triangle / or
"invalid" to tell the triangle is not valid

Rules
- sum of the lengths of the two shortest sides must be greater
than the length of the longest side
- every side must have a length greater than 0 
- if any of the above two condition is not satisfied, invalid 
triangle
Questions

Data Structures
- int, float 
- if all three sides have the same length, return 'E'
- two sides have the same length, third one is different, return 'I'
- all three sides different: return 'S'

Algorithms:
IDEA
- guard clauses:
    - if any of the number is 0, return invalid
    - if the sum of two shorter lenghts aren't greater than the 
    longest one, return invalid 

Pseudocode:
- nums = [num1, num2, num3]
- using list NUMS and list comprehension, check invalid
- using min and max func, determine shortest and longest
- determine middle_val from min and max
- if sum shorter lengths not > longest one, return invalid
- min == middle == max, return 'Equilateral'
- min == middle OR middle == max, return 'Isosceles'
- min != middle AND middle != max, return 'Scalene'
'''

# [1, 3, 4]
def triangle(num1, num2, num3):
    sides = [num1, num2, num3]
    if any([side <= 0 for side in sides]):
        return 'invalid'
    min_side = min(sides)
    max_side = max(sides)
    sides.remove(min_side)
    sides.remove(max_side)
    middle_side = sides[0]
    if not (min_side + middle_side > max_side):
        return 'invalid'
    if min_side == middle_side == max_side:
        return 'equilateral'
    elif (min_side == middle_side) or (middle_side == max_side):
        return 'isosceles'
    elif (min_side != middle_side) and (middle_side != max_side):
        return 'scalene'
    

print(triangle(3, 3, 3) == "equilateral")  # True
print(triangle(3, 3, 1.5) == "isosceles")  # True
print(triangle(3, 4, 5) == "scalene")      # True
print(triangle(0, 3, 3) == "invalid")      # True
print(triangle(3, 1, 1) == "invalid")      # True

File: exercises/LS215/medium2/triangle-sides.js
/*
Input: 3 triangle sides (Number, int)
Output: a string 
      - 3 valid types or 'invalid'

Rules:
- valid triangle:
  - sum of the lengths of the two shortest sides must be greater than the length of the longest side
  - every side must have a length greater than 0
  - "invalid" triangle

=== Data Structures ===
- number (int)
- array: Array.prototype.every method

=== Branstorm ===
- longest = Math.max(s1, s2, s3)
- shortest = Math.min(s1, s2, s3)
- middle = sum - (longest + shortest)
    => shortest + middle > longest 
- [s1, s2, s3].every(length all greater than 0)

=== Algorithms === 
*** isEquilateral (HELPER function) ***
- Input: 2 ints
  - max and min lengths
- Output: boolean
  - if (max === min), return true
  - otherwise, return false 

*** isScalene (HELPER function) *** 
- Input: 3 ints
- Output: boolean
  - if s1 !== s2 && s2 !== s3, return true
  - else, return false 

*** isInvalid (HELPER function) ***
- Input: 3 ints
- Output: boolean
  - if 3 sides consist invalid, return `true`
  - otherwise, return `false`

- if not shortest + middle > longest 
- if not [s1, s2, s3].every(length all greater than 0)

*** Main function (triangle) ***
- longest = Math.max(s1, s2, s3)
- shortest = Math.min(s1, s2, s3)
- middle = sum - (longest + shortest)
    => shortest + middle > longest 
- [s1, s2, s3].every(length all greater than 0)

- If 3 sides consist invalid triangle, return 'invalid'
  - HELPER (isInvalid)

- If 3 sides consist `Equilateral` (HELPER, isEquilateral)
  - return `equilateral`
- else if 3 sides consist `Scalene` (HELPER, isScalene)
  - return `scalene`
- else,
  - return `isosceles`
*/
function triangle(s1, s2, s3) {
  let sum = [s1, s2, s3].reduce((accm, num) => accm + num, 0);
  let longest = Math.max(s1, s2, s3);
  let shortest = Math.min(s1, s2, s3);
  let middle = sum - (longest + shortest);

  if (isInvalid(shortest, middle, longest)) {
    return 'invalid';
  }

  if (isEquilateral(shortest, longest)) {
    return 'equilateral';
  } else if (isScalene(shortest, middle, longest)) {
    return 'scalene';
  } else {
    return 'isosceles';
  }
}

function isEquilateral(min, max) {
  if (min === max) {
    return true;
  }

  return false;
}

function isScalene(s1, s2, s3) {
  if (s1 !== s2 && s2 !== s3) {
    return true;
  }

  return false;
}

function isInvalid(short, middle, long) {
  let sides = [short, middle, long];
  if (short + middle <= long) {
    return true;
  } else if (!sides.every(side => side > 0)) {
    return true;
  }

  return false;
}

console.log(triangle(3, 3, 3));        // "equilateral"
console.log(triangle(3, 3, 1.5));      // "isosceles"
console.log(triangle(3, 4, 5));        // "scalene"
console.log(triangle(0, 3, 3));        // "invalid"
console.log(triangle(3, 1, 1));        // "invalid"